### ISSUE TITLE
Metrics for test data

### ISSUE THREAD

**Issue by Abhinav28112003** [ultralytics/ultralytics#13599]
<img width="927" alt="Screenshot 2024-06-14 at 11 44 46 PM" src="https://github.com/ultralytics/ultralytics/assets/62950500/59cbf8bc-36fe-4ba4-a9dc-c2d8ac87c7d2">
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question
Hi @glenn-jocher
How do we get the metrics for test data? I have been seeing the past questions but I am not able to get the solution.
Do we need to do the labelling for test data as well?
And I don't know how to fine-tune my model like how to know how much IOU we need to give to the mode? Can you help me out with these questions? I am new to Yolo

**Comment by glenn-jocher** [comment]
Hello @Abhinav28112003,

Thank you for reaching out! Let's address your questions one by one:

### Metrics for Test Data
To evaluate your model on test data, you indeed need labeled test data. The labels allow the model to compare its predictions against the ground truth and compute metrics such as mAP, precision, recall, and F1 score. Here's a step-by-step guide:

1. **Label Your Test Data**: Ensure your test dataset is labeled in the same format as your training and validation datasets.
2. **Run Validation on Test Data**: You can use the `val` mode to evaluate your model on the test dataset. For example, if you have a YOLOv8 model and a labeled test dataset, you can run:
    ```bash
    yolo val model=path/to/your_model.pt data=path/to/test_data.yaml
    ```
   This command will output various performance metrics for your test data.

### Fine-Tuning Your Model
Fine-tuning involves adjusting various hyperparameters to improve model performance. Here are some tips:

1. **Intersection over Union (IoU)**: IoU is a crucial metric for object detection. To set an appropriate IoU threshold, you can start with a common value like 0.5 and adjust based on your specific needs. Higher IoU thresholds (e.g., 0.75) are stricter and may reduce false positives but might also miss some true positives. You can set the IoU threshold in the validation command:
    ```bash
    yolo val model=path/to/your_model.pt data=path/to/test_data.yaml iou=0.5
    ```

2. **Hyperparameter Tuning**: Experiment with different learning rates, batch sizes, and other hyperparameters. For example, to train with a different learning rate:
    ```bash
    yolo train data=path/to/your_data.yaml model=path/to/your_model.pt epochs=100 lr0=0.01
    ```

For a deeper dive into performance metrics and how to interpret them, you can refer to our [YOLO Performance Metrics Guide](https://docs.ultralytics.com/guides/yolo-performance-metrics/).

If you encounter any issues or need further assistance, please ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade your packages using:
```bash
pip install --upgrade torch ultralytics
```

Feel free to share any specific errors or issues you encounter, along with a minimum reproducible example, so we can assist you better. You can find guidelines for creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Happy detecting! 😊

**Comment by Abhinav28112003** [comment]
> Hello @Abhinav28112003,
> 
> Thank you for reaching out! Let's address your questions one by one:
> 
> ### Metrics for Test Data
> To evaluate your model on test data, you indeed need labeled test data. The labels allow the model to compare its predictions against the ground truth and compute metrics such as mAP, precision, recall, and F1 score. Here's a step-by-step guide:
> 
> 1. **Label Your Test Data**: Ensure your test dataset is labeled in the same format as your training and validation datasets.
> 2. **Run Validation on Test Data**: You can use the `val` mode to evaluate your model on the test dataset. For example, if you have a YOLOv8 model and a labeled test dataset, you can run:
>    ```shell
>    yolo val model=path/to/your_model.pt data=path/to/test_data.yaml
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    This command will output various performance metrics for your test data.
> 
> ### Fine-Tuning Your Model
> Fine-tuning involves adjusting various hyperparameters to improve model performance. Here are some tips:
> 
> 1. **Intersection over Union (IoU)**: IoU is a crucial metric for object detection. To set an appropriate IoU threshold, you can start with a common value like 0.5 and adjust based on your specific needs. Higher IoU thresholds (e.g., 0.75) are stricter and may reduce false positives but might also miss some true positives. You can set the IoU threshold in the validation command:
>    ```shell
>    yolo val model=path/to/your_model.pt data=path/to/test_data.yaml iou=0.5
>    ```
> 2. **Hyperparameter Tuning**: Experiment with different learning rates, batch sizes, and other hyperparameters. For example, to train with a different learning rate:
>    ```shell
>    yolo train data=path/to/your_data.yaml model=path/to/your_model.pt epochs=100 lr0=0.01
>    ```
> 
> For a deeper dive into performance metrics and how to interpret them, you can refer to our [YOLO Performance Metrics Guide](https://docs.ultralytics.com/guides/yolo-performance-metrics/).
> 
> If you encounter any issues or need further assistance, please ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade your packages using:
> 
> ```shell
> pip install --upgrade torch ultralytics
> ```
> 
> Feel free to share any specific errors or issues you encounter, along with a minimum reproducible example, so we can assist you better. You can find guidelines for creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Happy detecting! 😊

Thank you for the response😄.
